lib/ansible/plugins/vars/__init__.py

```diff
@@ -38,4 +43,6 @@ def __init__(self):
 
     def get_vars(self, loader, path, entities):
         """ Gets variables. """
-        self._basedir = basedir(path)
+        if path not in CANONICAL_PATHS:
+            CANONICAL_PATHS[path] = os.path.realpath(basedir(path))
+        self._basedir = CANONICAL_PATHS[path]
```

Comment:
I always thought this overall pattern was weird, (smuggling call state as instance data), and at least for the current usage, for no apparent benefit. Caching aside, this `self._basedir` set technically invalidates the `stateless` thing. So maybe `host_group_vars` shouldn't call this via `super()`, and this cache should just be moved inline to `host_group_vars.get_vars` and both should use a local instead of `self._basedir`, or an instance method that provides access to the cache (since a `super()` call can't just change the `path` value to the `realpath`, which I think is really what we want for this case?).

Reply:
'basedir' is mostly 'ported over' from when vars plugins were hardcoded plugins inside inventory and vars manager code. I always meant to clean it up, but was never a priority.

Reply:
I've just left `basedir` as-is for now, but I moved the cache for the realpath into host_group_vars, and it's no longer calling super().get_vars().